File: services/terrain_follower.py
"""
Terrain-following mode: adjusts waypoint altitudes to maintain constant AGL.
Adds intermediate waypoints along segments where terrain changes significantly.
"""
from services.elevation import get_elevations, get_path_elevations


def apply_terrain_following(waypoints, target_agl_m=30, interpolation_samples=5):
    """Adjust waypoints to maintain constant AGL above terrain.

    Args:
        waypoints: list of waypoint dicts
        target_agl_m: desired height above ground level
        interpolation_samples: intermediate points per segment

    Returns:
        list of adjusted waypoint dicts (may include interpolated points)
    """
    if not waypoints:
        return []

    # Get elevation at each original waypoint
    coords = [(w["lat"], w["lng"]) for w in waypoints]
    elevations = get_elevations(coords)

    adjusted = []
    wp_index = 0

    for i, (wp, ground_elev) in enumerate(zip(waypoints, elevations)):
        ground_elev = ground_elev or 0

        # Add the original waypoint with adjusted altitude
        adj = dict(wp)
        adj["index"] = wp_index
        adj["altitude_m"] = ground_elev + target_agl_m
        adj["ground_elevation_m"] = ground_elev
        adj["agl_m"] = target_agl_m
        adjusted.append(adj)
        wp_index += 1

        # Add intermediate points between this and next waypoint
        if i < len(waypoints) - 1 and interpolation_samples > 0:
            next_wp = waypoints[i + 1]
            path_data = get_path_elevations(
                wp["lat"], wp["lng"],
                next_wp["lat"], next_wp["lng"],
                num_samples=interpolation_samples + 2,
            )

            # Skip first and last (they are the original waypoints)
            for pd in path_data[1:-1]:
                interp = dict(wp)
                interp["index"] = wp_index
                interp["lat"] = pd["lat"]
                interp["lng"] = pd["lng"]
                interp["altitude_m"] = (pd["elevation"] or 0) + target_agl_m
                interp["ground_elevation_m"] = pd["elevation"] or 0
                interp["agl_m"] = target_agl_m
                interp["action_type"] = None  # No action at intermediate points
                interp["hover_time_s"] = 0.0
                adjusted.append(interp)
                wp_index += 1

    return adjusted

File: php/app/Services/TerrainFollower.php
<?php

namespace App\Services;

/**
 * Terrain-following mode: adjusts waypoint altitudes to maintain constant AGL.
 * Ported from services/terrain_follower.py.
 */
class TerrainFollower
{
    /**
     * Adjust waypoints to maintain constant AGL above terrain.
     */
    public static function applyTerrainFollowing(
        array $waypoints, float $targetAglM = 30, int $interpolationSamples = 5
    ): array {
        if (empty($waypoints)) {
            return [];
        }

        $coords = array_map(fn($w) => [$w['lat'], $w['lng']], $waypoints);
        $elevations = Elevation::getElevations($coords);

        $adjusted = [];
        $wpIndex = 0;

        foreach ($waypoints as $i => $wp) {
            $groundElev = $elevations[$i] ?? 0;

            $adj = $wp;
            $adj['index'] = $wpIndex;
            $adj['altitude_m'] = $groundElev + $targetAglM;
            $adj['ground_elevation_m'] = $groundElev;
            $adj['agl_m'] = $targetAglM;
            $adjusted[] = $adj;
            $wpIndex++;

            // Add intermediate points between this and next waypoint
            if ($i < count($waypoints) - 1 && $interpolationSamples > 0) {
                $nextWp = $waypoints[$i + 1];
                $pathData = Elevation::getPathElevations(
                    $wp['lat'], $wp['lng'],
                    $nextWp['lat'], $nextWp['lng'],
                    $interpolationSamples + 2
                );

                // Skip first and last (they are the original waypoints)
                $middle = array_slice($pathData, 1, -1);
                foreach ($middle as $pd) {
                    $interp = $wp;
                    $interp['index'] = $wpIndex;
                    $interp['lat'] = $pd['lat'];
                    $interp['lng'] = $pd['lng'];
                    $interp['altitude_m'] = ($pd['elevation'] ?? 0) + $targetAglM;
                    $interp['ground_elevation_m'] = $pd['elevation'] ?? 0;
                    $interp['agl_m'] = $targetAglM;
                    $interp['action_type'] = null;
                    $interp['hover_time_s'] = 0.0;
                    $adjusted[] = $interp;
                    $wpIndex++;
                }
            }
        }

        return $adjusted;
    }
}
